### ISSUE TITLE
FilePattern cannot load classpath resource from within jar file

### ISSUE THREAD

**Issue by devkanro** [jOOQ/jOOQ#11637]
@lukaseder  Maybe we should not discuss this issue in a closed PR #11613, so I create an issue for it.

### Description
In my case, we publish Jar that contains XML database schema to our maven repository.
Then we use those Jar for our code generating.
I got `URI is not hierarchical` error in code generating, I have debugged it and I am sure this is a bug of JOOQ.

We can see [line 227](https://github.com/jOOQ/jOOQ/blob/50b7ba5e76cab02ccfe4a04895546f106b5d6393/jOOQ/src/main/java/org/jooq/FilePattern.java#L227) of `jOOQ/src/main/java/org/jooq/FilePattern.java`
```Java
load0(new File(url.toURI()), loader);
```
The `url` is set by [line 214](https://github.com/jOOQ/jOOQ/blob/50b7ba5e76cab02ccfe4a04895546f106b5d6393/jOOQ/src/main/java/org/jooq/FilePattern.java#L214)
```Java
url = FilePattern.class.getResource(pattern);
```
As we know, create a `File` from classpath resource URI will cause the `URI is not hierarchical` error.
See the discussion in
https://stackoverflow.com/questions/10144210/java-jar-file-use-resource-errors-uri-is-not-hierarchical

### Steps to reproduce the problem
1. Just add an XML file as a resource to a Jar.
2. Add that Jar to the generator classpath.
3. Use the resource in XMLDatabase.
4. Run the generator. 

I also create an example to reproduce the problem.
See the [jooq-bugs.zip](https://github.com/jOOQ/jOOQ/files/6132339/jooq-bugs.zip), run the `run.bat` or execute this command
```Shell
java -cp jooq-3.14.8.jar;jooq-meta-3.14.8.jar;jooq-codegen-3.14. 8.jar;reactive-streams-1.0.2.jar;schema.jar org.jooq.codegen.GenerationTool configuration.xml
```
You will get the error:
```log
org.jooq.codegen.GeneratorException: Error generating code for catalog
        at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:441)
        at org.jooq.codegen.GenerationTool.run0(GenerationTool.java:879)
        at org.jooq.codegen.GenerationTool.run(GenerationTool.java:233)
        at org.jooq.codegen.GenerationTool.generate(GenerationTool.java:228)
        at org.jooq.codegen.GenerationTool.main(GenerationTool.java:200)
Caused by: java.lang.RuntimeException: java.lang.RuntimeException: Error while opening files /schema.xml or null
        at org.jooq.meta.AbstractDatabase.onError(AbstractDatabase.java:3104)
        at org.jooq.meta.AbstractDatabase.getSchemata(AbstractDatabase.java:647)
        at org.jooq.meta.AbstractDatabase.getSchemata(AbstractDatabase.java:679)
        at org.jooq.meta.CatalogDefinition.getSchemata(CatalogDefinition.java:63)
        at org.jooq.codegen.JavaGenerator.generateCatalogIfEmpty(JavaGenerator.java:476)
        at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:435)
        ... 4 more
Caused by: java.lang.RuntimeException: Error while opening files /schema.xml or null
        at org.jooq.meta.xml.XMLDatabase.info(XMLDatabase.java:239)
        at org.jooq.meta.xml.XMLDatabase.getSchemata0(XMLDatabase.java:506)
        at org.jooq.meta.AbstractDatabase$3.run(AbstractDatabase.java:650)
        at org.jooq.meta.AbstractDatabase.onError(AbstractDatabase.java:3094)
        ... 9 more
Caused by: java.lang.IllegalArgumentException: URI is not hierarchical
        at java.io.File.<init>(File.java:418)
        at org.jooq.FilePattern.load(FilePattern.java:225)
        at org.jooq.meta.xml.XMLDatabase.info(XMLDatabase.java:164)
        ... 12 more
```

### Versions

- jOOQ: 3.14.8
- Java: 8

**Comment by lukaseder** [comment]
Thank you very much for your patience. I'm sorry for the delay with this issue. It seems that the notification has been lost.

I can reproduce the exception from your example ZIP file. However, I don't agree with the fix. I had much rather call `Class::getResourceAsStream` instead of `URL::openStream`.

Will look into a fix, now

**Comment by lukaseder** [comment]
Also, given the API method `FilePattern.collect()`, we cannot keep an open `InputStream` in a resulting `Source`, so we have to eagerly read the content in that case.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0, 3.16.1 (https://github.com/jOOQ/jOOQ/issues/12790), and 3.15.6 (https://github.com/jOOQ/jOOQ/issues/12791)

Thanks again for your report!
